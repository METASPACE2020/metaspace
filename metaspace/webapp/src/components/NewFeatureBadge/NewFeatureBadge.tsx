import { reactive, ref } from 'vue'
import { ElBadge } from '../../lib/element-plus'
import { defineComponent } from 'vue'
// Import the CSS as needed
import './NewFeatureBadge.css'

// Assuming getLocalStorage and setLocalStorage are compatible with Vue 3
import { getLocalStorage, setLocalStorage } from '../../lib/localStorage'

const storageKey = 'new_feature_badges'

const store = reactive<any>(getLocalStorage(storageKey) || {})

export function hideFeatureBadge(featureKey) {
  if (store[featureKey]) {
    return
  }
  store[featureKey] = true
  setLocalStorage(storageKey, store)
}

export default defineComponent({
  name: 'NewFeatureBadge',
  components: {
    ElBadge,
  },
  props: {
    customClass: { type: String, default: '' },
    featureKey: { type: String, required: true },
    showUntil: { type: Date, required: false },
  },
  setup(props, { slots }) {
    const isStale = ref(props.showUntil && props.showUntil.valueOf() < Date.now())

    if (!(props.featureKey in store)) {
      store[props.featureKey] = false
    }

    return () => {
      if (store[props.featureKey] || isStale.value) {
        return slots.default ? slots.default() : null
      }
      return (
        <ElBadge
          value="New"
          isDot
          class={['sm-feature-badge', { 'sm-feature-badge--hidden': store[props.featureKey] }, props.customClass]}
        >
          {slots.default ? slots.default() : null}
        </ElBadge>
      )
    }
  },
})
