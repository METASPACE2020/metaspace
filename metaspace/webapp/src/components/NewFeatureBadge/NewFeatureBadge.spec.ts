// Import necessary Vue 3 and Vuex 4 functionalities
import { defineComponent, nextTick, h } from 'vue'
import { mount } from '@vue/test-utils'
import NewFeatureBadge, { hideFeatureBadge } from './NewFeatureBadge'

const TestNewFeatureBadge = defineComponent({
  name: 'TestNewFeatureBadge',
  props: {
    featureKey: String,
  },
  setup(props) {
    return () =>
      h(
        NewFeatureBadge,
        { ...props },
        {
          default: () => h('span', {}, 'Badge Content'),
          // Include other named slots if necessary
        }
      )
  },
})

const featureKey = 'test'

describe('NewFeatureBadge', () => {
  it('should render a badge', async () => {
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should hide the badge', async () => {
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    hideFeatureBadge(featureKey)
    await nextTick()
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should not render the badge if already hidden', () => {
    hideFeatureBadge(featureKey)
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should not render the badge if stale', () => {
    const originalDateNow = Date.now
    Date.now = vi.fn(() => new Date('2020-01-02T00:00:00.000').valueOf())

    const wrapper = mount(TestNewFeatureBadge, {
      props: { featureKey, showUntil: new Date('2020-01-01T00:00:00.000') },
    })

    expect(wrapper.classes()).toMatchSnapshot()

    // Restore the original Date.now() function
    Date.now = originalDateNow
  })
})
