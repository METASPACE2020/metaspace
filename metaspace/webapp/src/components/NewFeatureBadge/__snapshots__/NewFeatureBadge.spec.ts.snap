// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewFeatureBadge > should hide the badge 1`] = `[]`;

exports[`NewFeatureBadge > should not render the badge if already hidden 1`] = `[]`;

exports[`NewFeatureBadge > should not render the badge if stale 1`] = `[]`;

exports[`NewFeatureBadge > should render a badge 1`] = `
[
  "el-badge",
  "sm-feature-badge",
]
`;
